### PR TITLE
Feat/top 5 merchant by revenue

### LINKED ISSUE #32: Admin Mercahnts: Top 5 Merchants by Revenue
Admin Merchants: Top 5 Merchants by Revenue

As an admin,
When I visit the admin merchants index
Then I see the names of the top 5 merchants by total revenue generated
And I see that each merchant name links to the admin merchant show page for that merchant
And I see the total revenue generated next to each merchant name

Notes on Revenue Calculation:
- Only invoices with at least one successful transaction should count towards revenue
- Revenue for an invoice should be calculated as the sum of the revenue of all invoice items
- Revenue for an invoice item should be calculated as the invoice item unit price multiplied by the quantity (do not use the item unit price)

### PR DESCRIPTION
**Pull request by Dominicod**
Solo
* Created tests for admin merchant top 5 by revenue
* Created logic for admin merchant top 5 by revenue
* Fixed failing tests due to ID association

I believe my AR query is doing what is required, however what is currently tripping me up is the story in bold.

Notes on Revenue Calculation:
- Only invoices with at least one successful transaction should count towards revenue
**- Revenue for an invoice should be calculated as the sum of the revenue of all invoice items**
- Revenue for an invoice item should be calculated as the invoice item unit price multiplied by the quantity (do not use the item unit price)

### CODE REVIEW

**Review by lcole37 — APPROVED** [review]

File: app/models/customer.rb [hunk 1]
```diff
@@ -20,6 +20,11 @@ def self.top_five_customers(merchant_id)
   end
 
   def self.top_five_customers_admin
-    joins(:transactions).where('transactions.result = 0').group(:last_name, :first_name).order(count: :desc).limit(5).count
+    joins(:transactions)
+      .where('transactions.result = 0')
+      .group(:last_name, :first_name)
+      .order(count: :desc)
+      .limit(5)
+      .count
```
Comment (rebeckahendricks) [inline app/models/customer.rb:28]: I am pretty sure this is not going to give us what you want... I am also not sure about this, but I have had more luck with using the aggregate in my select statement and providing an alias (that way you can use it later in the view).  Like this: `.select('invoice_items.*, sum('invoice_items.quantity * invoice_items.unit_price) as revenue')` and I also think you should include ':first_name, :last_name, in your select statement but then `.group(:id)` and `.order('revenue')`

Comment (rebeckahendricks) [reply app/models/customer.rb:28]: Oh wait, I was using the user story from your PR request... This is a different method. I see it below haha. Ignore most of that - but I would still think about a select statement here that might do more of what you want it to.

File: app/models/merchant.rb [hunk 2]
```diff
@@ -16,10 +16,19 @@ def disabled_items
...
   end
 
   def self.disabled_merchants
-    all.where(status: "disabled")
+    where(status: "disabled")
+  end
+
+  def self.top_five_by_revenue
+    joins(:items, invoices: :transactions)
+      .select('merchants.*, sum(invoice_items.unit_price * invoice_items.quantity) as revenue')
```
Comment (rebeckahendricks) [inline app/models/merchant.rb:28]: You said you weren't sure about this part:  Revenue for an invoice should be calculated as the sum of the revenue of all invoice items

But I do beleive that your aggregate in your select statement does this.

**Review by rebeckahendricks — APPROVED** [review]
I don't see any glaring issues, but you definitely will want to find a way to test this to be certain that it is working.

**Review by thayes87 — APPROVED** [review]

**Pull request merged**